ethcore/src/snapshot/consensus/authority.rs

```diff
@@ -413,84 +332,57 @@ impl Rebuilder for ChunkRebuilder {
 				}
 			}
 
-			let last_hashes: Vec<H256> = last_rlp.list_at(4)?;
-			let parent_td: ::util::U256 = last_rlp.val_at(5)?;
+			let parent_td: ::util::U256 = last_rlp.val_at(4)?;
 
 			let mut batch = self.db.transaction();
 			self.chain.insert_unordered_block(&mut batch, &block_data, receipts, Some(parent_td), true, false);
 			self.db.write_buffered(batch);
 
-			self.warp_target = Some((block.header, last_hashes));
+			self.warp_target = Some(block.header);
 		}
 
 		Ok(())
 	}
 
-	fn finalize(&mut self, db: StateDB, engine: &Engine) -> Result<(), ::error::Error> {
-		use state::State;
-
+	fn finalize(&mut self, _engine: &Engine) -> Result<(), ::error::Error> {
 		if !self.had_genesis {
 			return Err(Error::WrongChunkFormat("No genesis transition included.".into()).into());
 		}
 
-		let (target_header, target_last_hashes) = match self.warp_target.take() {
+		let target_header = match self.warp_target.take() {
 			Some(x) => x,
 			None => return Err(Error::WrongChunkFormat("Warp target block not included.".into()).into()),
 		};
 
-		// we store the last data even for the last chunk for easier verification
-		// of warp target, but we don't store genesis transition data.
-		// other than that, there should be a one-to-one correspondence of
-		// chunk ends to chunk beginnings.
-		if self.last_proofs.len() != self.unverified_firsts.len() + 1 {
-			return Err(Error::WrongChunkFormat("More than one 'last' chunk".into()).into());
-		}
-
 		// verify the first entries of chunks we couldn't before.
-		let lasts_iter = self.last_proofs.iter().map(|&(_, ref hdr, ref proof)| (hdr, &proof[..]));
-		let firsts_iter = self.unverified_firsts.iter().map(|&(_, ref hdr)| hdr);
+		// we store all last verifiers, but not all firsts.
+		// match each unverified first epoch with a last epoch verifier.
+		let mut lasts_reversed = self.last_epochs.iter().rev();
```

Comment:
a reversed `[T]` iterator is just pointer-bumping, same as an in-order iterator. The loop itself just walks two vectors in reverse, which should be pretty fast.